src/window/mod.rs

```diff
@@ -1,852 +1,600 @@
 
-impl From<WindowState> for bool {
-    fn from(s: WindowState) -> bool {
-        match s {
-            WindowState::Active => true,
-            WindowState::Inactive => false,
-        }
-    }
-}
+pub(crate) mod inner;
 
-/// Possible events generated by a window that you need to handle
-#[derive(Clone, Debug)]
-pub enum Event {
-    /// The state of your window has been changed
-    Configure {
-        /// Optional new size for your *inner* surface
-        ///
-        /// This is the new size of the contents of your window
-        /// as suggested by the server. You can ignore it and choose
-        /// a new size if you want better control on the possible
-        /// sizes of your window.
-        ///
-        /// The size is expressed in logical pixels, you need to multiply it by
-        /// your buffer scale to get the actual number of pixels to draw.
-        ///
-        /// In all cases, these events can be generated in large batches
-        /// during an interactive resize, and you should buffer them before
-        /// processing them. You only need to handle the last one of a batch.
-        new_size: Option<(u32, u32)>,
-        /// New combination of states of your window
-        ///
-        /// Typically tells you if your surface is active/inactive, maximized,
-        /// etc...
-        states: Vec<State>,
-    },
-    /// A close request has been received
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DecorationMode {
+    /// The server should decide which decoration mode to use.
     ///
-    /// Most likely the user has clicked on the close button of the decorations
-    /// or something equivalent
-    Close,
-    /// The decorations need to be refreshed
-    Refresh,
-}
+    /// This is probably the best option.
+    ServerDecides,
 
-/// Possible decoration modes for a Window
-///
-/// This represents what your application requests from the server.
-///
-/// In any case, the compositor may override your requests. In that case SCTK
-/// will follow its decision.
-///
-/// If you don't care about it, you should use `FollowServer` (which is the
-/// SCTK default). It'd be the most ergonomic for your users.
-#[derive(Debug, PartialEq, Eq)]
-pub enum Decorations {
-    /// Request server-side decorations
+    /// Requests that the server should draw decorations if possible.
     ServerSide,
-    /// Force using the client-side `Frame`
+
+    /// The client will draw decorations.
     ClientSide,
-    /// Follow the preference of the compositor
-    FollowServer,
-    /// Don't decorate the Window
+
+    /// The client should not have any decorations.
     None,
 }
 
-struct WindowInner<F> {
-    frame: Rc<RefCell<F>>,
-    shell_surface: Arc<Box<dyn shell::ShellSurface>>,
-    user_impl: Box<dyn FnMut(Event, DispatchData)>,
-    min_size: (u32, u32),
-    max_size: Option<(u32, u32)>,
-    current_size: (u32, u32),
-    old_size: Option<(u32, u32)>,
-    decorated: bool,
+pub trait ShellHandler<D> {
```

Comment:
I see this module is still very WIP, but overall it'd be nice if we manage to closely fit the API of libdecor, in terms of what shell-related methods are available, and what shell events are forwarded to the user. It'll make future integration with libdecor much easier.

For reference, the public API of libdecor is here: https://gitlab.gnome.org/jadahl/libdecor/-/blob/master/src/libdecor.h